### ISSUE TITLE
Poor error messages when forgetting to 'extern module x' in ci files

### ISSUE THREAD

**Issue by epmikida** [UIUC-PPL/charm#1101]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/1101*

---
If you have a setup where you have two modules, m1 and m2, and m1 creates chares declared in m2 but you have forgotten to add 'extern module m2;' inside of module m1, the code will compile but segfault in a variety of ways with no useful indication of why. It won't be caught at compile time, but at runtime there should be more useful error handling.

A simple example where this occurs is below. Since the mainmodule m1 does not extern module m2 this code compiles, but segfaults when run on more than 1 PE (so far only tested in SMP mode). The code for this example is also attached. 

Examples of errors also pasted below. The error varied on a run to run basis. The command used to produce all errors was './charmrun ++local +p8 ++ppn 2 error' on a 16 core machine. The build was a net linux build, sometimes with production, sometimes without as shown below.

*In the ci file:*
mainmodule m1 {
  mainchare Main {
    entry Main(CkArgMsg*);
  };
};

module m2 {
  group Grp {
    entry Grp();
  };
};

*In the .C file:*
Main::Main(CkArgMsg* msg) {
  CProxy_Grp::ckNew();
}

*Examples of Errors:*
Charmrun> error on request socket to node 0 '127.0.0.1'--
Socket closed before recv.
^^ Shows up both --with-production and without

------------- Processor 8 Exiting: Caught Signal ------------
Reason: segmentation violation
Suggestion: Try running with '++debug', or linking with '-memory paranoid' (memory paranoid requires '+netpoll' at runtime).
[8] Stack Traceback:
  [8:0]   [0x50c0a7]
  [8:1] +0x35670  [0x7fe66f21b670]
  [8:2] TransmitImplicitDgram+0x19  [0x50a5e9]
  [8:3] TransmitDatagram+0x18f  [0x50aa0f]
  [8:4]   [0x50d42f]
  [8:5] CommunicationServerThread+0xe  [0x50dfae]
  [8:6]   [0x50e3ab]
  [8:7]   [0x50e425]
  [8:8] +0x7dc5  [0x7fe66ffd4dc5]
  [8:9] clone+0x6d  [0x7fe66f2dcced]
Fatal error on PE 8> segmentation violation
^^ Only seems to show up --with-production

Fatal error on PE 11> [11] Fatal error #4559312
^^ Only seems to show up --with-production

------------- Processor 8 Exiting: Called CmiAbort ------------
Reason: CmiFree reference count was zero-- is this a duplicate free?
[8] Stack Traceback:
  [8:0] CmiAbortHelper+0xcb  [0x5ef5d4]
  [8:1] CmiAbort+0x2f  [0x5ef611]
  [8:2] CmiFree+0x3c  [0x5fbb0b]
  [8:3] GarbageCollectMsg+0x27  [0x5f2362]
  [8:4] DiscardImplicitDgram+0x35  [0x5f23a5]
  [8:5] IntegrateAckDatagram+0x306  [0x5f39dc]
  [8:6] ReceiveDatagram+0x213  [0x5f3c3e]
  [8:7]   [0x5f3e24]
  [8:8] LrtsAdvanceCommunication+0x1a  [0x5f434d]
  [8:9]   [0x5ef325]
  [8:10]   [0x5ef33c]
  [8:11] CommunicationServerThread+0x15  [0x5ef38c]
  [8:12]   [0x5ef2b6]
  [8:13]   [0x5ec1f4]
  [8:14] +0x7dc5  [0x7f5941c81dc5]
  [8:15] clone+0x6d  [0x7f5940f89ced]
^^ Only seems to showup without --with-production

[error.ci](https://github.com/UIUC-PPL/charm/files/3536596/error.ci.txt)
[Main.C](https://github.com/UIUC-PPL/charm/files/3536597/Main.C.txt)

**Comment by epmikida** [comment]
*Original date: 2016-06-14 21:39:08*

---
Here's the relevant slack conversation:

Eric Mikida
FWIW, if you have multiple charm modules, and don't use extern module X; in your main module then it still compiles, and runs, but produces wildly different and unhelpful errors.

Sanjay Kale [3:17 PM]  
hmm.. I thought this was caught in part by checking for “0” as unregistered EPs. I don’t know if that error message in the code.
Of course, that depends on if unitialized variables are 0

Nikhil Jain [3:17 PM]  
I think it crashes with saying unregistered something
like group or chare

Sanjay Kale [3:17 PM]  
That shouldbe fine. Maybe we can enhance the error message to state probable cause.
Unregistered xyz. Has every module (or the module containing this EP/chare/group) been included via a chain of "extern module” constructs from the main module.
But we can strengthen it (not just relying on 0 values) by having a static in variable (such as _ck_module_initialized ) and assigning to it some values thats difficult to arise by chance. version number?
Then we have to check (maybe only in debug mode), if this variable has the correct value from some code in that file.. hmm.. something in defs.h? or each unmarshalling function.. (when compiled under debug flag). 
`mikida: what are some of the “wildly different errors”?

Eric Mikida [3:36 PM]  
i thought it crashed with those errors too which is why it took me so long to find the issue. ive forgotten to extern modules before,  but this time i got errors like the following:

the most common error looks like this:

 ```------------- Processor 4 Exiting: Caught Signal ------------
Reason: segmentation violation
Suggestion: Try running with '++debug', or linking with '-memory paranoid' (memory paranoid requires '+netpoll' at runtime).
[4] Stack Traceback:
  [4:0]   [0x515227]
  [4:1] +0x35670  [0x7f96d03a3670]
  [4:2] TransmitImplicitDgram1+0x19  [0x513849]
  [4:3] TransmitDatagram+0x138  [0x513b38]
  [4:4]   [0x5165af]
  [4:5] CommunicationServerThread+0xe  [0x51712e]
  [4:6]   [0x51752b]
  [4:7]   [0x5175a5]
  [4:8] +0x7dc5  [0x7f96d115cdc5]
  [4:9] clone+0x6d  [0x7f96d0464ced]```

Nikhil Jain [3:43 PM]  
this is weird stuff. crashing in network layer is the worst way to crash.

Eric Mikida [3:43 PM]  
yeah
heres another:

 ```------------- Processor 1 Exiting: Caught Signal ------------
Reason: segmentation violation
Charmrun> error on request socket to node 0 '127.0.0.1'--
Socket closed before recv.```

Nikhil Jain [3:44 PM]  
oh, this is my favorite
the only thing we learn from this crash is that we crashed.

Eric Mikida [3:44 PM]  
nothing in either of these gave me any indication that it was something as simple as forgetting an extern module

Nikhil Jain [3:45 PM]  
which is odd. i wonder if group registration checking etc are CkAsserts and u r running with production

Eric Mikida [3:45 PM]  
yes i believe i am running with production. i dont have direct access to the build, but ill try the debug build

Nikhil Jain [3:46 PM]  
yeah, try that and see if a ckasserts catches it, in which case, we go back to Jim’s old complain - a CkAssert != assert.
and in any case, an assert is not also the right way to catch such erorrs in my opinion. if checking with ckabort is the best.

Eric Mikida [3:47 PM]  
still sometimes crashes via "socket closed before recv"
but the other error now shows up as:

Reason: CmiFree reference count was zero-- is this a duplicate free?
 
still not the most helpful, but better than nothing

Nikhil Jain [3:47 PM]  
yeah, all this means the error was not caught when it actualluy happened, and now we are seeing its effects
there isn;t much we can do when socket closed before recv” happens.
bcoz i dont think it lets us handle anything at that pooint.

Eric Mikida [3:48 PM]  
im mainly just confused about what is different now then when in the past where ive seen the more helpful error message

Nikhil Jain [3:49 PM]  
production vs non production is where i put money on
or to spell out, most likely ckassert is being used to test the eror conditions that threw the helpful error message in the past. but when building with production, ckassert is defined to nothing. and hence lets the error pass.

Eric Mikida [3:51 PM]  
im pretty sure when i build with disneys debug build of charm that they would not have with-production on, but ill find out.

Nikhil Jain [3:51 PM]  
oh, u tried that?
i dint realize u changed builds.

Eric Mikida [3:53 PM]  
yea. changing builds here is very easy
i just dont know how the builds were created in the first place.
but i assume the double free warning was a CkAssert that is now showing up with the debug build.

Phil Miller [4:27 PM]  
You can check the build configuration in `conv-autoconfig.h`. Also, do we not do something to assign EP0 to an error handler? That would be zero overhead, and I think catch almost all instances of this error

Sanjay Kale [4:31 PM]  
i am pretty sure we do.. that is in the ancient DNA of the code.
But I suspect its crashing before any message to EPs in that module are sent.. die to ther unregistered stuff?
But worth getting to the bottom of this.

**Comment by epmikida** [comment]
*Original date: 2016-06-14 21:40:11*

---
TL;DR of the above conversation to summarize: Various people seem to remember that we handled this more gracefully by having EP0 be an error handler or something along those lines, but apparently this is no longer the case.
